api/v1alpha1/gateway_types.go

```diff
@@ -179,8 +178,6 @@ type ListenerTLS struct {
 
 // GatewayStatus defines the observed state of Gateway
 type GatewayStatus struct {
-	// TODO overall status
```

Comment:
I think it's important that overall `GatewayStatus` is defined in this PR to gain a full understanding of status. I'm using `Pod` status as a reference model and it feels like we currently have the scope of status reversed. Instead of surfacing conditions in the subresources, maybe conditions should be in`GatewayStatus`. For example:
```
status:
  conditions:
  - lastTransitionTime: "2020-01-27T17:13:49Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2020-01-27T17:13:49Z"
    status: "True"
    type: ListenersReady
  - lastTransitionTime: "2020-01-27T17:13:10Z"
    status: "True"
    type: GatewayScheduled
  listenerStatuses:
    - <listener[0] status>
  routeStatuses:
    - <route[0] status>
```
Should `ListenerStatus` and `GatewayRouteStatus` follow a similar approach to [ContainerStatus](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2189-L2207)? For example: 

```
// RouteStatus represents the status of a route
type RouteStatus struct {
	// Each route in a namespace must have a unique name.
	NamespacedName types.NamespacedName
	// +optional
        // This is where ConditionInvalidRoute and ConditionNoSuchRoute can be used.
	State RouteState
	// Should a route have some type of readiness check?
	Ready bool
	// +optional
	RouteID string
}

// ListenerStatus represents the status of a listener.
type ListenerStatus struct {
	// Address bound on this listener.
	Address *ListenerAddress `json:"address"`
        // State reflects if the listener is good or bad (bad address, port, protocol, cert ref, etc..)
        // This is where ConditionInvalidAddress etc. can be used.
	// +optional
	State ListenerState
	// Should a listener have some type of readiness check?
	Ready bool
}
```

Reply:
I agree with the general idea of having a status.Conditions set that indicates if there are problems with listeners and routes, but I think that, following the API conventions, they should probably be `ListenersNotReady`, `RoutesNotReady`, `GatewayClassInvalid`, and `GatewayNotScheduled`.

Reply:
This is great feedback, thanks! I think my latest update incorporates most of this, but let me know if I've missed anything.

Reply:
Note that for things that are continuously reconciled, the "well-known" types from Deployments and ReplicaSets (and stateful sets and daemonsets) should be considered - primarily "Progressing" and "Available" (available may be more relevant here).

In general the pattern of `kubectl wait --for=available` is specifically designed to deal with the problem of "how do I know when my changes have been made available after an update" in concert with `ObservedGeneration`.  I would recommend at least having an `Available` condition that matches the requested semantics (when an ingress controller has accepted or begun filling traffic) and `ObservedGeneration int64` set by an ingress controller to indicate that the requested spec has been observed and that status conditions apply to that generation of the ingress.

Reply:
I like the idea of reusing these types, as well as any other relevant ones.